src/windows/common/hcs_schema.h

```diff
@@ -481,13 +495,135 @@ struct ComputeSystem
     NLOHMANN_DEFINE_TYPE_INTRUSIVE_ONLY_SERIALIZE(ComputeSystem, Owner, ShouldTerminateOnLastHandleClosed, SchemaVersion, VirtualMachine)
 };
 
+struct GuestErrorSaveReport
+{
+    std::optional<std::wstring> SaveStateFile;
+    std::optional<long> Status;
+};
+
+inline void to_json(nlohmann::json& j, const GuestErrorSaveReport& g)
+{
+    j = nlohmann::json::object();
+    OMIT_IF_EMPTY(j, g, SaveStateFile);
+    OMIT_IF_EMPTY(j, g, Status);
+}
+
+inline void from_json(const nlohmann::json& j, GuestErrorSaveReport& r)
+{
+    if (j.contains("SaveStateFile"))
+        r.SaveStateFile = j.at("SaveStateFile").get<std::wstring>();
+    if (j.contains("Status"))
+        r.Status = j.at("Status").get<int32_t>();
+}
+
 struct CrashReport
 {
     std::wstring CrashLog;
+    std::optional<GuestErrorSaveReport> GuestCrashSaveInfo;
+};
+
+inline void to_json(nlohmann::json& j, const CrashReport& c)
+{
+    j = nlohmann::json::object();
+    j.at("CrashLog") = c.CrashLog;
+    OMIT_IF_EMPTY(j, c, GuestCrashSaveInfo);
+}
 
-    NLOHMANN_DEFINE_TYPE_INTRUSIVE_WITH_DEFAULT(CrashReport, CrashLog);
+inline void from_json(const nlohmann::json& j, CrashReport& c)
+{
+    if (j.contains("CrashLog"))
+        c.CrashLog = j.at("CrashLog").get<std::wstring>();
+    if (j.contains("GuestCrashSaveInfo"))
+        c.GuestCrashSaveInfo = j.at("GuestCrashSaveInfo").get<GuestErrorSaveReport>();
+}
+
+enum class NotificationType
+{
+    None,
+    GracefulExit,
+    ForcedExit,
+    UnexpectedExit,
+    Unknown
 };
 
+NLOHMANN_JSON_SERIALIZE_ENUM(
+    NotificationType,
+    {
+        {NotificationType::None, "None"},
+        {NotificationType::GracefulExit, "GracefulExit"},
+        {NotificationType::ForcedExit, "ForcedExit"},
+        {NotificationType::UnexpectedExit, "UnexpectedExit"},
+        {NotificationType::Unknown, "Unknown"},
+    })
+
+struct GuestCrashAttribution
+{
+    std::optional<std::vector<uint64_t>> CrashParameters;
+};
+
+inline void to_json(nlohmann::json& j, const GuestCrashAttribution& g)
+{
+    j = nlohmann::json::object();
+    if (g.CrashParameters.has_value())
```

Comment:
I do not see this pattern elsewhere in the file, @OneBlue knows this stuff better than me, there is probably an easier way.

Reply:
On the to_json direction that's the best way I found. 

There's probably a smarter way to make nlohmann_json handle std::optional<T> for us, but that's probably something that we'll want as a separate change

Reply:
It should be possible rewrite those as: 

```
OMIT_IF_EMPTY(j, GuestCrashAttribution, CrashParameters):
```

though

Reply:
I would prefer we use the OMIT_IF_EMPTY macro like the rest of the code does.